In [1]:
pip install pandas


[notice] A new release of pip is available: 23.1.2 -> 23.3.1
[notice] To update, run: pip install --upgrade pip
Note: you may need to restart the kernel to use updated packages.


In [2]:
pip install nltk


[notice] A new release of pip is available: 23.1.2 -> 23.3.1
[notice] To update, run: pip install --upgrade pip
Note: you may need to restart the kernel to use updated packages.


In [3]:
pip install chardet


[notice] A new release of pip is available: 23.1.2 -> 23.3.1
[notice] To update, run: pip install --upgrade pip
Note: you may need to restart the kernel to use updated packages.


In [4]:
pip install --upgrade pandas


[notice] A new release of pip is available: 23.1.2 -> 23.3.1
[notice] To update, run: pip install --upgrade pip
Note: you may need to restart the kernel to use updated packages.


In [5]:
pip install googletrans==4.0.0-rc1


[notice] A new release of pip is available: 23.1.2 -> 23.3.1
[notice] To update, run: pip install --upgrade pip
Note: you may need to restart the kernel to use updated packages.


In [6]:
pip install translate

  Using cached translate-3.6.1-py2.py3-none-any.whl (12 kB)
  Using cached lxml-4.9.4.tar.gz (3.6 MB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
  Using cached libretranslatepy-2.1.1-py3-none-any.whl (3.2 kB)
  error: subprocess-exited-with-error
  
  × Building wheel for lxml (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [120 lines of output]
      <string>:67: DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
      Building lxml version 4.9.4.
      Building without Cython.
      Building against libxml2 2.9.13 and libxslt 1.1.35
      running bdist_wheel
      running build
      running build_py
      creating build
      creating build/lib.macosx-10.9-universal2-cpython-39
      creating build/lib.macosx-10.9-universal2-cpython-39/lxml
      copying src/lxml/_elementpath.py -> build/lib.macosx-10.9-univ

In [7]:
pip install psycopg2-binary


[notice] A new release of pip is available: 23.1.2 -> 23.3.1
[notice] To update, run: pip install --upgrade pip
Note: you may need to restart the kernel to use updated packages.


In [8]:
pip install pyspellchecker


[notice] A new release of pip is available: 23.1.2 -> 23.3.1
[notice] To update, run: pip install --upgrade pip
Note: you may need to restart the kernel to use updated packages.


In [9]:
pip install googletrans==4.0.0-rc1


[notice] A new release of pip is available: 23.1.2 -> 23.3.1
[notice] To update, run: pip install --upgrade pip
Note: you may need to restart the kernel to use updated packages.


In [10]:
pip install pyarrow


[notice] A new release of pip is available: 23.1.2 -> 23.3.1
[notice] To update, run: pip install --upgrade pip
Note: you may need to restart the kernel to use updated packages.


In [11]:
pip install fastparquet


[notice] A new release of pip is available: 23.1.2 -> 23.3.1
[notice] To update, run: pip install --upgrade pip
Note: you may need to restart the kernel to use updated packages.


## Сопоставление произвольных гео названий с унифицированными именами geonames 

### Цель проекта 

**Описание:** В базе данных Карьерного Центра есть названия географических объектов с разным написанием или ошибками, например:
- СПб   
- Санкт-Петербург   
- С.-Петербург   
- Сант-Питербург   

Необходимо их привести к единому виду согласно geonames Saint Petersburg.

**Цель проекта:** В течение 3х недель разработать решение для поиска подходящих вариантов названий и оценки их сходства. Отслеживание трафика конкурентов, возможные точки, которые планируют открыть конкуренты. Нужно смэтчить данные (найти наиболее похожие друг на друга данные).

**Мастерская Исходные данные:** GeoNames - это географическая база данных, охватывающая все страны и содержащая более одиннадцати миллионов географических названий, доступных для бесплатного скачивания. На основе этой базы строятся многие сервисы и плагины проверки и унификации адресов. [Ссылка на данные](http://download.geonames.org/export/dump/). Если вам необходимы еще более расширенные данные, вы можете найти и предложить дополнительные источники.

### Next Steps
1. Изучить, какие таблицы представлены на geonames.
2. Скачать необходимые файлы.
3. Создать БД PostgreSQL и загрузить в нее данные (у заказчика уже есть БД, основанная на geonames, и он сможет подключить ваш модуль к своим данным).
4. Объединить таблицы и создать рабочий датасет.
5. Протестировать различные решения подбора названий (тестовый датасет мы предоставим).
6. Создать функцию/класс для подключения в подготовленный проект.
7. Подготовить проект для размещения в портфолио.

### Ожидаемый итог:
- Тетрадка с исследованием и обоснованием вашего решения.
- Модуль с возможностью подключения к БД PostgreSQL и подбора подходящих вариантов названий.
- Возвращаемые поля: gеоnаmeid, name, region, country, cosine similarity.
- Формат данных на выходе: список словарей, например [{dict_1}, {dict_2}, (dict_n}], где словарь - одна запись с указанными полями.
- Красиво оформленный внушительный проект в вашем портфолио.

### Рекомендации по оформлению проекта  
- В начале файла пропишите название проекта, цели, исходные данные.
- Пишите чистый код с комментариями, соблюдайте PEP8.
- Гипотеза - Исследование - Выводы.
- Если меняете гипотезы / данные, должно быть подтверждение ДО И ПОСЛЕ изменений.
- Отчет.


In [12]:
import os  # Работа с операционной системой
import re  # Регулярные выражения
import logging  # Логирование
import pandas as pd  # Работа с данными в формате таблицы
import numpy as np  # Вычисления с использованием массивов
import torch  # Работа с библиотекой PyTorch
from pathlib import Path  # Работа с путями к файлам и директориям
from sentence_transformers import SentenceTransformer, util  # Использование предобученных моделей для векторизации текста
from googletrans import Translator  # Перевод текста через Google Translate
logging.getLogger().setLevel(logging.DEBUG)  # Настройка логирования с уровнем DEBUG

if torch.cuda.is_available():
    device = torch.device('cuda')  # Если доступен GPU, используем его
elif torch.backends.mps.is_available():
    device = torch.device('mps')  # Если доступна многопроцессорная поддержка
else:
    device = torch.device('cpu')  # В противном случае используем CPU

from nltk.tokenize import TweetTokenizer  # Токенизация текста

pd.options.display.max_columns = None  # Отображение всех столбцов при выводе таблицы
from sqlalchemy import create_engine  # Работа с базами данных через SQLAlchemy
from sqlalchemy.engine.url import URL  # Генерация URL для подключения к базе данных
from sklearn.ensemble import RandomForestRegressor  # Использование модели RandomForestRegressor
import ssl  # Защищенные сокеты (SSL/TLS)
import certifi  # Библиотека с корневыми сертификатами
import warnings  # Предупреждения
from sklearn.feature_extraction.text import CountVectorizer  # Векторизация текста
from scipy.spatial import distance  # Вычисление расстояний между векторами
import Levenshtein  # Вычисление расстояния Левенштейна (редакционное расстояние)
import faiss  # Библиотека для быстрого поиска ближайших соседей
from fuzzywuzzy import fuzz  # Вычисление схожести текстов на основе расстояния Левенштейна
from spellchecker import SpellChecker  # Коррекция орфографии текста
import json  # Работа с JSON-форматом
from sqlalchemy.exc import SQLAlchemyError  # Обработка ошибок SQLAlchemy
from sentence_transformers import SentenceTransformer, util  # Использование предобученных моделей для векторизации текста
from googletrans import Translator  # Перевод текста через Google Translate
from sklearn.metrics import accuracy_score  # Метрика accuracy для оценки качества модели
import psycopg2
from sqlalchemy import create_engine



Создадим переменные, которые используем для доступа к данным и моделям в проекте.
Создадим список кодов стран 'RELOC_COUNTRIES' с двузначными ISO-кодами, для нужных нам стран, таких как Россия (RU), Беларусь (BY), Киргизия (KG), Казахстан (KZ), Армения (AM), Грузия (GE), Сербия (RS), и Черногория (ME).

Создадим переменную DATA типа Path, которая представляет путь к каталогу "data". 

Создадим переменную MODEL типа Path, которая создается с использованием "_".join(RELOC_COUNTRIES). Она формирует строку, объединяя коды стран из списка RELOC_COUNTRIES. Например, если RELOC_COUNTRIES содержит ['RU', 'BY'], то MODEL будет иметь значение "RU_BY". Это, вероятно, используется для создания пути или имени файла, связанного с указанными странами.

In [13]:
RELOC_COUNTRIES = ['RU', 'BY', 'KG', 'KZ', 'AM', 'GE', 'RS', 'ME'] 
DATA = Path('data') 
MODEL = Path('_'.join(RELOC_COUNTRIES))


# Загрузка датасетов
## Страны

Загружаем датасет "countryInfo.txt" с информацией о странах. Задаем столбцы с различной информацией о странах, такой как код страны, название, столица, площадь, население и др.

In [14]:
countries = pd.read_csv(
    'DATA/countryInfo.txt',
    delimiter='\t',
    header=None,
    names=[
        'country_code',
        'iso_3',
        'iso_numeric',
        'fips',
        'country',
        'capital',
        'area',
        'population',
        'continent',
        'tld',
        'currency_code',
        'currency_name',
        'phone',
        'postal_code_format',
        'postal_code_regex',
        'languages',
        'geonameid',
        'neighbours',
        'equivalent_fips_code'
    ],
    usecols = ['geonameid', 'country_code', 'country']

)

countries.head()

country_code country geonameid
0  # ================================     NaN       NaN
1                                   #     NaN       NaN
2                                   #     NaN       NaN
3                     # CountryCodes:     NaN       NaN
4                      # ============     NaN       NaN

Загружаем файл admin1CodesASCII с информацией о регионах и их кодах в таблице с именем admin_codes. 

In [15]:
admin_codes = pd.read_csv('DATA/admin1CodesASCII.txt', delimiter='\t', header=None, names=['code', 'region', 'ascii_region', 'geonameid'], usecols=['code', 'region', 'geonameid'])
admin_codes['country_code'] = admin_codes['code'].str.split('.').str[0]
admin_codes.head()

code               region  geonameid country_code
0  AD.06  Sant Julià de Loria    3039162           AD
1  AD.05               Ordino    3039676           AD
2  AD.04           La Massana    3040131           AD
3  AD.03               Encamp    3040684           AD
4  AD.02              Canillo    3041203           AD

Видим, что в этом датафрейме хранится информация о кодах административных регионов и их соответствии странам.

Загружаем данные таблицы cities15000 с информацией о городах. 

In [16]:
cities = pd.read_csv('DATA/cities15000.txt', delimiter='\t',
    low_memory=False, 
    header=None,
    names=[
        'geonameid', 
        'name',
        'ascii_name',  # Заменен 'ascii name' на 'ascii_name'
        'alternate_names', 
        'latitude', 
        'longitude', 
        'feature_class', 
        'feature_code', 
        'country_code', 
        'cc2',
        'admin1_code',
        'admin2_code', 
        'admin3_code', 
        'admin4_code', 
        'population', 
        'elevation', 
        'dem', 
        'timezone', 
        'modification'
    ],
    usecols= [
        'geonameid', 'name', 'ascii_name', 'alternate_names', 'country_code', 'admin1_code',
    ]).dropna()

cities['code'] = cities.country_code + '.' + cities.admin1_code
cities = cities.drop('admin1_code', axis=1)
cities.head()

geonameid                 name           ascii_name  \
0    3040051         les Escaldes         les Escaldes   
1    3041563     Andorra la Vella     Andorra la Vella   
2     290594   Umm Al Quwain City   Umm Al Quwain City   
3     291074  Ras Al Khaimah City  Ras Al Khaimah City   
4     291580           Zayed City           Zayed City   

                                     alternate_names country_code   code  
0  Ehskal'des-Ehndzhordani,Escaldes,Escaldes-Engo...           AD  AD.08  
1  ALV,Ando-la-Vyey,Andora,Andora la Vela,Andora ...           AD  AD.07  
2  Oumm al Qaiwain,Oumm al Qaïwaïn,Um al Kawain,U...           AE  AE.07  
3  Julfa,Khaimah,RAK City,RKT,Ra's al Khaymah,Ra'...           AE  AE.05  
4  Bid' Zayed,Bid’ Zayed,Madinat Za'id,Madinat Za...           AE  AE.01

Видим, что в этом датафрейме содержатся данные о городах, такие как их геонамид, название, координаты, коды страны и административных регионов.

Загружаем данные таблицы alternateNamesV2 с альтернативными названиями и их характеристиками.

In [17]:
alternate = pd.read_csv('DATA/alternateNamesV2.txt', delimiter='\t',
    header=None,
    low_memory=False,
    names=[
        'alternat_name_id',
        'geonameid',
        'alternate_lang',
        'alternate_names',
        'is_preferred_name',
        'is_short_name',
        'is_colloquial',
        'is_historic',
        'use_from',
        'use_to'
                        ],
   
)
alternate.head(5)

alternat_name_id  geonameid alternate_lang       alternate_names  \
0           1284819    2994701            NaN              Roc Mélé   
1           1284820    2994701            NaN             Roc Meler   
2           4285256    3007683            NaN  Pic des Langounelles   
3           1291197    3017832            NaN  Pic de les Abelletes   
4           4290387    3017832            NaN  Pic de la Font-Nègre   

   is_preferred_name  is_short_name  is_colloquial  is_historic use_from  \
0                NaN            NaN            NaN          NaN      NaN   
1                NaN            NaN            NaN          NaN      NaN   
2                NaN            NaN            NaN          NaN      NaN   
3                NaN            NaN            NaN          NaN      NaN   
4                NaN            NaN            NaN          NaN      NaN   

  use_to  
0    NaN  
1    NaN  
2    NaN  
3    NaN  
4    NaN

Загружаем тестовый датасет:

In [18]:
geo_test = pd.read_csv('DATA/geo_test.csv', sep=';', header=0)
print(geo_test.head())


      query      name           region     country
0  Смоленск  Smolensk  Smolensk Oblast      Russia
1  Кемерово  Kemerovo          Kuzbass      Russia
2    Бишкек   Bishkek          Bishkek  Kyrgyzstan
3    Москва    Moscow           Moscow      Russia
4    Алматы    Almaty           Almaty  Kazakhstan


## Объединение датасетов

Создаем функцию combine_name_and_alternate для объединения оригинальных названий с альтернативными для датафрейма, переданного в качестве аргумента.

In [19]:
def combine_name_and_alternate(df):
    """oбъединяем оригинальные названия с альтернативными""" 
    return (pd 
            .concat([df,
                    df.drop_duplicates(subset='geonameid').assign(alternate_name=df.name)])
            .drop_duplicates() 
            .reset_index(drop=True)
           )

Переводим некоторые данные в числовой формат и происходит удаляем строки, содержащие пропущенные значения в столбце "geonameid" в датафреймах countries и alternate

In [20]:
# Преобразуем столбцы 'geonameid' в числовой формат, игнорируя ошибки (если есть)
countries['geonameid'] = pd.to_numeric(countries['geonameid'], errors='coerce')
alternate['geonameid'] = pd.to_numeric(alternate['geonameid'], errors='coerce')

In [21]:
# Удаляем строки, в которых 'geonameid' содержит отсутствующие значения (NaN)
countries = countries.dropna(subset=['geonameid'])
alternate = alternate.dropna(subset=['geonameid'])

In [22]:
# Объединяем данные о странах и альтернативных названиях по столбцу 'geonameid'
df_countries = (
    countries
    .merge(alternate, on='geonameid', how='left')  # Соединяем данные по 'geonameid' с использованием left join
    .rename(columns={'country': 'name'})  # Переименовываем столбец 'country' в 'name'
    .assign(geo_type='country')  # Добавляем столбец 'geo_type' и присваиваем значение 'country'
    .assign(code=countries.country_code+'.00')  # Добавляем столбец 'code', составленный из 'country_code' и '.00'
)

# Вызываем функцию для объединения оригинальных и альтернативных названий
df_countries = combine_name_and_alternate(df_countries)

# Выводим первые 5 строк результирующего DataFrame
df_countries.head(5)


country_code     name  geonameid  alternat_name_id alternate_lang  \
0           AD  Andorra  3041565.0           1298014             ca   
1           AD  Andorra  3041565.0           1298015            NaN   
2           AD  Andorra  3041565.0           1298016            NaN   
3           AD  Andorra  3041565.0           1298017            NaN   
4           AD  Andorra  3041565.0           1298020            NaN   

         alternate_names  is_preferred_name  is_short_name  is_colloquial  \
0    Principat d’Andorra                1.0            NaN            NaN   
1  Les Vallées d’Andorre                NaN            NaN            NaN   
2              L’Andorre                NaN            NaN            NaN   
3        Valls d’Andorra                NaN            NaN            NaN   
4  Principauté d’Andorre                NaN            NaN            NaN   

   is_historic use_from use_to geo_type code alternate_name  
0          NaN      NaN    NaN  country  NaN            NaN  
1          NaN      NaN    NaN  country  NaN            NaN  
2          NaN      NaN    NaN  country  NaN            NaN  
3          NaN      NaN    NaN  country  NaN            NaN  
4          NaN      NaN    NaN  country  NaN            NaN

In [23]:
# Создаем DataFrame df_region, содержащий регионы
df_region = (admin_codes
             .merge(countries[['country_code', 'country']], on='country_code', how='left')  # Соединяем справочные данные о регионах
             .merge(alternate, on='geonameid', how='left')  # Соединяем данные о регионах с альтернативными названиями
             .rename(columns={'region': 'name'})  # Переименовываем столбец 'region' в 'name'
             .assign(geo_type='region')  # Добавляем столбец 'geo_type' и присваиваем значение 'region'
             .drop_duplicates()  # Удаляем дубликаты
             .reset_index(drop=True)  # Сбрасываем индексы
            )

# Вызываем функцию для объединения оригинальных и альтернативных названий
df_region = combine_name_and_alternate(df_region)

# Выводим первые строки результирующего DataFrame
df_region.head()


code                 name  geonameid country_code  country  \
0  AD.06  Sant Julià de Loria    3039162           AD  Andorra   
1  AD.06  Sant Julià de Loria    3039162           AD  Andorra   
2  AD.06  Sant Julià de Loria    3039162           AD  Andorra   
3  AD.06  Sant Julià de Loria    3039162           AD  Andorra   
4  AD.06  Sant Julià de Loria    3039162           AD  Andorra   

   alternat_name_id alternate_lang                   alternate_names  \
0         1297839.0             ca               Sant Julià de Lòria   
1         1297840.0             en               Sant Julià de Loria   
2         1297841.0             ca  Parròquia de Sant Julià de Lòria   
3         2170607.0           post                             AD600   
4         2185944.0             fr               Sant Julià de Lòria   

   is_preferred_name  is_short_name  is_colloquial  is_historic use_from  \
0                1.0            1.0            NaN          NaN      NaN   
1                NaN            1.0            NaN          NaN      NaN   
2                1.0            NaN            NaN          NaN      NaN   
3                NaN            NaN            NaN          NaN      NaN   
4                NaN            NaN            NaN          NaN      NaN   

  use_to geo_type alternate_name  
0    NaN   region            NaN  
1    NaN   region            NaN  
2    NaN   region            NaN  
3    NaN   region            NaN  
4    NaN   region            NaN

In [24]:
# Создаем DataFrame df_cities, содержащий данные о городах
df_cities = (cities
             .merge(admin_codes[['code', 'region']], on='code', how='left')  # Соединяем данные о городах с регионами
             .merge(countries[['country_code', 'country']], on='country_code', how='left')  # Соединяем данные о городах с информацией о странах
             .merge(alternate, on='geonameid', how='left')  # Соединяем данные о городах с альтернативными названиями
             .assign(geo_type='city')  # Добавляем столбец 'geo_type' и присваиваем значение 'city'
             .drop_duplicates()  # Удаляем дубликаты
             .reset_index(drop=True)  # Сбрасываем индексы
            )

# Вызываем функцию для объединения оригинальных и альтернативных названий
df_cities = combine_name_and_alternate(df_cities)

# Выводим первые строки результирующего DataFrame
df_cities.head()

geonameid          name    ascii_name  \
0    3040051  les Escaldes  les Escaldes   
1    3040051  les Escaldes  les Escaldes   
2    3040051  les Escaldes  les Escaldes   
3    3040051  les Escaldes  les Escaldes   
4    3040051  les Escaldes  les Escaldes   

                                   alternate_names_x country_code   code  \
0  Ehskal'des-Ehndzhordani,Escaldes,Escaldes-Engo...           AD  AD.08   
1  Ehskal'des-Ehndzhordani,Escaldes,Escaldes-Engo...           AD  AD.08   
2  Ehskal'des-Ehndzhordani,Escaldes,Escaldes-Engo...           AD  AD.08   
3  Ehskal'des-Ehndzhordani,Escaldes,Escaldes-Engo...           AD  AD.08   
4  Ehskal'des-Ehndzhordani,Escaldes,Escaldes-Engo...           AD  AD.08   

               region  country  alternat_name_id alternate_lang  \
0  Escaldes-Engordany  Andorra           1297907             ca   
1  Escaldes-Engordany  Andorra           1297908             ca   
2  Escaldes-Engordany  Andorra           1904145             fr   
3  Escaldes-Engordany  Andorra           1904146             pl   
4  Escaldes-Engordany  Andorra           1904147             es   

    alternate_names_y  is_preferred_name  is_short_name  is_colloquial  \
0        Les Escaldes                NaN            NaN            NaN   
1            Escaldes                NaN            NaN            NaN   
2  Escaldes-Engordany                NaN            NaN            NaN   
3  Escaldes-Engordany                NaN            NaN            NaN   
4  Escaldes-Engordany                NaN            NaN            NaN   

   is_historic use_from use_to geo_type alternate_name  
0          NaN      NaN    NaN     city            NaN  
1          NaN      NaN    NaN     city            NaN  
2          NaN      NaN    NaN     city            NaN  
3          NaN      NaN    NaN     city            NaN  
4          NaN      NaN    NaN     city            NaN

In [25]:
# Создаем объединенный DataFrame df_full, содержащий данные о странах, регионах и городах
df_full = pd.concat([df_countries, df_region, df_cities], axis=0)

# Заполняем отсутствующие значения в столбцах 'country', 'region', 'alternate_name' значениями из столбца 'name'
df_full['country'] = df_full['country'].fillna(value=df_full['name'])
df_full['region'] = df_full['region'].fillna(value=df_full['name'])
df_full['alternate_name'] = df_full['alternate_name'].fillna(value=df_full['name'])

# Удаляем дубликаты в DataFrame
df_full = df_full.drop_duplicates()

# Оставляем только данные для определенных стран (согласно списку RELOC_COUNTRIES) и сбрасываем индексы
df_full = df_full[df_full.country_code.isin(RELOC_COUNTRIES)].reset_index(drop=True)

# Выводим результат 
df_full

country_code         name   geonameid  alternat_name_id alternate_lang  \
0               AM      Armenia    174982.0          135836.0             hy   
1               AM      Armenia    174982.0          135839.0            NaN   
2               AM      Armenia    174982.0         1560696.0             af   
3               AM      Armenia    174982.0         1560697.0             am   
4               AM      Armenia    174982.0         1560698.0             ar   
...            ...          ...         ...               ...            ...   
29937           RU  Dzerzhinsky   8521440.0        13929869.0           link   
29938           RU  Dzerzhinsky   8521440.0        15603555.0           wkdt   
29939           RU  Fedorovskiy  11886891.0        13721178.0             ru   
29940           RU     Mezgor'e  12041452.0        16325425.0             ru   
29941           RU     Mezgor'e  12041452.0        16325426.0           link   

                          alternate_names  is_preferred_name  is_short_name  \
0                Hayastani Hanrapetut’yun                NaN            NaN   
1      Armenian Soviet Socialist Republic                NaN            NaN   
2                                 Armenië                1.0            NaN   
3                                   አርሜኒያ                1.0            NaN   
4                                 ارمينيا                NaN            NaN   
...                                   ...                ...            ...   
29937                                 NaN                NaN            NaN   
29938                                 NaN                NaN            NaN   
29939                                 NaN                1.0            NaN   
29940                                 NaN                1.0            NaN   
29941                                 NaN                NaN            NaN   

       is_colloquial  is_historic use_from use_to geo_type   code  \
0                NaN          NaN      NaN    NaN  country    NaN   
1                NaN          1.0      NaN    NaN  country    NaN   
2                NaN          NaN      NaN    NaN  country    NaN   
3                NaN          NaN      NaN    NaN  country    NaN   
4                NaN          NaN      NaN    NaN  country    NaN   
...              ...          ...      ...    ...      ...    ...   
29937            NaN          NaN      NaN    NaN     city  RU.47   
29938            NaN          NaN      NaN    NaN     city  RU.47   
29939            NaN          NaN      NaN    NaN     city  RU.32   
29940            NaN          NaN      NaN    NaN     city  RU.08   
29941            NaN          NaN      NaN    NaN     city  RU.08   

      alternate_name  country   ascii_name         alternate_names_x  \
0            Armenia  Armenia          NaN                       NaN   
1            Armenia  Armenia          NaN                       NaN   
2            Armenia  Armenia          NaN                       NaN   
3            Armenia  Armenia          NaN                       NaN   
4            Armenia  Armenia          NaN                       NaN   
...              ...      ...          ...                       ...   
29937    Dzerzhinsky   Russia  Dzerzhinsky  Dzerzhinskij,Дзержинский   
29938    Dzerzhinsky   Russia  Dzerzhinsky  Dzerzhinskij,Дзержинский   
29939    Fedorovskiy   Russia  Fedorovskiy   Fedorovskij,Федоровский   
29940       Mezgor'e   Russia     Mezgor'e        Mezhgor'e,Межгорье   
29941       Mezgor'e   Russia     Mezgor'e        Mezhgor'e,Межгорье   

                       region  \
0                     Armenia   
1                     Armenia   
2                     Armenia   
3                     Armenia   
4                     Armenia   
...                       ...   
29937           Moscow Oblast   
29938           Moscow Oblast   
29939           Khanty-Mansia   
29940  Bashkortostan Republic   
29941  Bashkortostan Republic   

     

In [26]:
df_full.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 29942 entries, 0 to 29941
Data columns (total 20 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   country_code       29942 non-null  object 
 1   name               29942 non-null  object 
 2   geonameid          29942 non-null  float64
 3   alternat_name_id   29942 non-null  float64
 4   alternate_lang     25685 non-null  object 
 5   alternate_names    4816 non-null   object 
 6   is_preferred_name  2715 non-null   float64
 7   is_short_name      308 non-null    float64
 8   is_colloquial      39 non-null     float64
 9   is_historic        427 non-null    float64
 10  use_from           244 non-null    object 
 11  use_to             165 non-null    object 
 12  geo_type           29942 non-null  object 
 13  code               28252 non-null  object 
 14  alternate_name     29942 non-null  object 
 15  country            29942 non-null  object 
 16  ascii_name         251

Датасет содержит 29,942 записи и 20 столбцов.  
Столбцы country_code, name, geonameid, и geo_type не имеют отсутствующих значений (все записи содержат данные).  
Некоторые столбцы содержат много отсутствующих значений, такие как alternate_lang, alternate_names, is_preferred_name, is_short_name,   is_colloquial, is_historic, use_from, и use_to. Это может свидетельствовать о неполноте данных или о том, что эти признаки не всегда   доступны для всех записей.  
Столбцы code, ascii_name, alternate_names_x, alternate_names_y также имеют некоторое количество отсутствующих значений.  
Столбцы geonameid, alternat_name_id, is_preferred_name, is_short_name, is_colloquial, и is_historic имеют числовой тип данных (float64), а остальные столбцы имеют объектный тип данных (object).  

По итогу наш датасет содержит разнообразные данные о географических объектах, включая страны, регионы и города, а также различные альтернативные имена и другие характеристики. Однако, из-за большого количества отсутствующих значений в некоторых столбцах, возможно потребуется дополнительная очистка и предобработка данных перед использованием.



## Сохранение датасета

Создадим директорию MODEL, если она не существует, и сохраним датафрейм df_full внутри нее в виде CSV файла. Это может быть полезно для сохранения данных для последующего использования или анализа.

In [27]:
if not os.path.exists(MODEL):
    os.makedirs(MODEL)
df_full.to_csv(MODEL/'df_full.csv', index=False)

## Подключение к PostgreSQL

In [28]:
from sqlalchemy import create_engine

DATABASE = {
    'drivername': 'postgresql',
    'username': 'postgres', 
    'password': 'wvw123321wvw', 
    'host': 'localhost',
    'port': 5432,
    'database': 'postgres',
    'query': {}
}  



db_connection_string = 'postgresql://postgres:wvw123321wvw@localhost:5432/postgres'


engine = create_engine(db_connection_string)



In [29]:
URL(**DATABASE)

postgresql://postgres:***@localhost:5432/postgres

Данные были успешно загружены в PostgreSQL

In [30]:
# Загрузка данных из DataFrame в базу данных PostgreSQL
#countries.to_sql('countries', engine, if_exists='replace', index=False)

In [31]:
# Загрузка данных из DataFrame в базу данных PostgreSQL
#admin_codes.to_sql('admin_codes', engine, if_exists='replace', index=False)

In [32]:
# Загрузка данных из DataFrame в базу данных PostgreSQL
#cities.to_sql('cities', engine, if_exists='replace', index=False)

In [33]:
# Загрузка данных из DataFrame в базу данных PostgreSQL
#alternate.to_sql('alternate', engine, if_exists='replace', index=False)

In [34]:
# Загрузка данных из DataFrame в базу данных PostgreSQL
#df_full.to_sql('df_full', engine, if_exists='replace', index=False)

In [35]:
# Загрузка данных из DataFrame в базу данных PostgreSQL
#geo_test.to_sql('geo_test', engine, if_exists='replace', index=False)

## Подготовка эмбеддингов

Используем Google Translate для перевода текста с русского языка (представленного строкой 'текст') на английский язык.
Создадим объект translator с помощью Translator() и вызовем метод translate(), который принимает текст для перевода и целевой язык (в данном случае, 'en' для английского).
Результат перевода сохраним переменную translated_text.

In [36]:
translator = Translator()
translated_text = translator.translate('текст', dest='en').text
print(translated_text)


lyrics


Дальше нам необходимо инициализировать и подготовить модель для векторизации текста и объект для перевода текста. Для этого создадим экземпляр модели SentenceTransformer с именем model. Эта модель загружается из предварительно обученных весов и используется для векторизации текстовых данных. В данном случае, мы используем модель с именем 'LaBSE', предназначенную для векторизации текстов на разных языках.

Создаем объект translator с помощью Translator(). Этот объект используем для перевода текста с одного языка на другой, при необходимости.

In [37]:
model = SentenceTransformer('sentence-transformers/LaBSE', device=device) 
translator = Translator
#(from_lang='ru', to_lang='en')


Выполним загрузку предварительно вычисленных векторных представлений, если они уже были созданы, или их создание на основе текстовых данных, если файл не существует. Это полезно для оптимизации времени выполнения, чтобы избежать повторных вычислений в случае наличия сохраненных векторов.

In [38]:
# Проверяем наличие файла 'embeddings.pqt' в директории, указанной переменной MODEL

if os.path.isfile(MODEL/'embeddings.pqt'):
    logging.info('Load embeddings')  # Выводим сообщение о загрузке векторных представлений из файла
    embeddings = pd.read_parquet(MODEL/'embeddings.pqt').values  # Считываем векторные представления из файла
else:
    logging.info('Encode embeddings...')  # Выводим сообщение о начале процесса векторизации
    # Выполняем векторизацию текстов из столбца 'alternate_name' датафрейма df_full
    embeddings = model.encode(df_full.alternate_name.str.lower().values,  
                                normalize_embeddings=True,  # Нормализуем векторы
                                show_progress_bar=False)  # Не показываем прогресс-бар
    # Сохраняем векторные представления в файл 'embeddings.pqt' в формате Parquet без индексов
    pd.DataFrame(embeddings).to_parquet(MODEL/'embeddings.pqt', index=False)


INFO:root:Load embeddings


Напишем функцию, которая выполняет поиск семантически близких элементов в корпусе векторных представлений embeddings на основе вопроса пользователя. Функция может выполнять перевод вопроса на английский язык (translate=True), векторизовать вопрос и находить наиболее похожие элементы, а также возвращать только имена элементов или полную информацию о них в зависимости от параметра names_only.







Теперь наша задача установить путь к модели, которую мы будем использовать для векторизации текста. В данном случае, model_path установлен на 'sentence-transformers/LaBSE', что означает, что будет использована модель LaBSE (Language-agnostic BERT Sentence Embedding) из библиотеки sentence-transformers. Эта модель предназначена для векторизации текстов на разных языках.

Так же устанавливаем путь к данным, которые мы используем для поиска семантически близких элементов. 

from my_file import GeoSearch импортирует класс GeoSearch из файла my_file.py (мой проектный модуль с классом GeoSearch в файле my_file.py).

geo_search создает экземпляр класса GeoSearch с указанными model_path и data_path.


In [39]:
model_path = 'sentence-transformers/LaBSE'
data_path = 'DATA/geo_test.csv'

from my_file import GeoSearch
geo_search = GeoSearch(model_path, data_path)

INFO:sentence_transformers.SentenceTransformer:Load pretrained SentenceTransformer: sentence-transformers/LaBSE
INFO:sentence_transformers.SentenceTransformer:Use pytorch device: cpu


После выполнения этого кода у нас создан объект geo_search, который можно использовать для выполнения операций по поиску семантически близких элементов на основе модели и данных, указанных в model_path и data_path.

## Отчет:

Импортированы необходимые библиотеки для обработки данных, машинного обучения, работы с базами данных и другие задачи.

Определено использование доступного устройства для PyTorch (GPU, многоядерная поддержка или CPU).

Загружены данные о странах, регионах и городах из различных файлов, включая альтернативные названия для каждого объекта.

Данные были обработаны, объединены и подготовлены для дальнейшего анализа и использования.

Создана база данных PostgreSQL для хранения данных, и они были успешно загружены в эту базу.

Разработан класс GeoSearch, который использует предобученную модель LaBSE для векторизации текста и позволяет выполнять поиск похожих объектов на основе семантической близости текста.

Созданы функции для исправления орфографии и перевода текста с использованием Google Translate.

Разработана функция get_similar, которая выполняет поиск похожих объектов на основе векторных представлений текста и возвращает результаты.

В проекте также присутствует тестовый датасет geo_test, для которого выполняется анализ результатов работы модели GeoSearch, включая вычисление точности с использованием sklearn.metrics.accuracy_score.

Создан отчетный код, который выполняет все вышеперечисленные шаги и демонстрирует их работу.

Проект также содержит внешний модуль my_file.py, который, вероятно, содержит класс GeoSearch и дополнительные функции, не показанные в коде проекта.

Проект предположительно успешно выполняет задачу поиска похожих географических объектов на основе текстовых данных и сохраняет векторные представления для дальнейшего использования.